src/desktop/imapimport/adsync/utils/DifferentialUidLoader.ts

```diff
@@ -0,0 +1,199 @@
+	private readonly importedUidToMailIdsMap: Map<number, ImapMailIds>
+	private readonly isEnableImapQresync: boolean
+	private readonly emitAdSyncEventTypes: Set<AdSyncEventType>
+
+	constructor(
+		imapClient: ImapFlow,
+		importedUidToMailIdsMap: Map<number, ImapMailIds>,
+		isEnableImapQresync: boolean,
+		emitAdSyncEventTypes: Set<AdSyncEventType>,
+	) {
+		this.imapClient = imapClient
+		this.importedUidToMailIdsMap = importedUidToMailIdsMap
+		this.isEnableImapQresync = isEnableImapQresync
+		this.emitAdSyncEventTypes = emitAdSyncEventTypes
+	}
+
+	async calculateUidDiff(lastFetchedMailSeq: number, downloadBatchSize: number, totalMessageCount: number | null): Promise<DeletedUids> {
+		this.uidDiffInProgress = true
+		// if IMAP QRESYNC is enabled and supported by the IMAP server, we do not need to calculate the diff on our own
+		if (this.isEnableImapQresync) {
+			this.uidFetchRequestQueue.push({
+				uidFetchSequenceString: "1:*",
+				fetchRequestType: UidFetchRequestType.QRESYNC,
+			})
+			this.uidDiffInProgress = false
+			return [] // delete events are handle automatically by IMAP QRESYNC extension
+		}
+
+		let seenUids = await this.calculateUidDiffInBatches(lastFetchedMailSeq + 1, downloadBatchSize, totalMessageCount)
+
+		let deletedUids: number[] = []
+		if (this.emitAdSyncEventTypes.has(AdSyncEventType.DELETE)) {
+			deletedUids = [...this.importedUidToMailIdsMap.keys()].filter((uid) => {
+				!seenUids.includes(uid)
+			})
+		}
+
+		this.uidDiffInProgress = false
+		return deletedUids
+	}
+
+	private async calculateUidDiffInBatches(fromSeq: number, downloadBatchSize: number, totalMessageCount: number | null): Promise<SeenUids> {
+		let seenUids: number[] = []
+
+		if (totalMessageCount == 0) {
+			return seenUids
+		}
+
+		let toSeq = fromSeq + downloadBatchSize
+		let isFinalBatch = totalMessageCount == null || toSeq > totalMessageCount
+
+		let fromSeqString = fromSeq.toString()
+		let toSeqString = totalMessageCount == null ? "*" : isFinalBatch ? totalMessageCount.toString() : toSeq.toString()
+
+		let mails = this.imapClient.fetch(`${fromSeqString}:${toSeqString}`, { uid: true }, {})
+
+		// @ts-ignore
+		for await (const mail of mails) {
+			let uid = mail.uid
+			seenUids.push(uid)
+
+			if (this.importedUidToMailIdsMap.has(uid)) {
+				this.uidUpdateQueue.push(uid)
+			} else {
+				this.uidCreateQueue.push(uid)
+			}
+		}
+
+		if (!isFinalBatch) {
+			seenUids.push(...(await this.calculateUidDiffInBatches(toSeq + 1, downloadBatchSize, totalMessageCount)))
+		}
+
+		return seenUids
+	}
+
+	async getNextUidFetchRequest(downloadBatchSize: number): Promise<UidFetchRequest | null> {
+		let waitingUidFetchRequest = this.uidFetchRequestQueue.pop()
+		if (waitingUidFetchRequest) {
+			return waitingUidFetchRequest
+		}
+
+		let nextUidFetchRange: number[]
+		let fetchRequestType: UidFetchRequestType
+
+		if (this.emitAdSyncEventTypes.has(AdSyncEventType.CREATE) && this.uidCreateQueue.length != 0) {
```

Comment:
I think this are flags that are populated from the config to turn on/off certain events